python/tests/udf/sql_test.py

```diff
@@ -0,0 +1,127 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+from keras.applications import InceptionV3
+from keras.applications import inception_v3 as iv3
+
+from pyspark import SparkContext
+from pyspark.sql import DataFrame, Row
+from pyspark.sql.functions import udf
+
+from sparkdl.udf.keras_image_model import registerKerasImageUDF
+from sparkdl.graph.builder import IsolatedSession
+import sparkdl.utils.jvmapi as JVMAPI
+from sparkdl.image.imageIO import imageSchema, imageArrayToStruct
+from ..tests import SparkDLTestCase
+from ..transformers.image_utils import getSampleImagePathsDF
+
+def get_image_paths_df(sqlCtx):
+    df = getSampleImagePathsDF(sqlCtx, "fpath")
+    df.createOrReplaceTempView("_test_image_paths_df")
+    return df
+
+
+class SqlUserDefinedFunctionTest(SparkDLTestCase):
+
+    def test_composite_udf(self):
+        df = get_image_paths_df(self.sql)
+        df.show()
+
+        def keras_load_img(fpath):
+            from keras.preprocessing.image import load_img, img_to_array
+            import numpy as np
+            from pyspark.sql import Row            
+            img = load_img(fpath, target_size=(299, 299))
+            return img_to_array(img).astype(np.uint8)
+
+        def pil_load_spimg(fpath):
+            from PIL import Image
+            import numpy as np
+            img_arr = np.array(Image.open(fpath), dtype=np.uint8)
+            return imageArrayToStruct(img_arr)
+
+        def keras_load_spimg(fpath):
+            return imageArrayToStruct(keras_load_img(fpath))
+    
+        # Load image with Keras and store it in our image schema
+        JVMAPI.registerUDF('keras_load_spimg', keras_load_spimg, imageSchema)        
+        JVMAPI.registerUDF('pil_load_spimg', pil_load_spimg, imageSchema)
+
+        # Register an InceptionV3 model
+        registerKerasImageUDF("iv3_img_pred",
+                              InceptionV3(weights="imagenet"),
+                              keras_load_img)
+
+        SQL = self.session.sql
+
+        # Choice 1: manually chain the functions in SQL
+        df1 = SQL("select iv3_img_pred(keras_load_spimg(fpath)) as preds from _test_image_paths_df")
+        preds1 = np.array(df1.select("preds").rdd.collect())
+
+        # Choice 2: build a pipelined UDF and directly use it in SQL
+        JVMAPI.registerPipeline("load_img_then_iv3_pred", ["keras_load_spimg", "iv3_img_pred"])
```

Comment:
one potential optimization:
For this use case where the input dataframe to the SQL UDF contains file paths, instead of having to register a pipeline of load => predict, we could have a version of registerKerasUDF that natively handles file path input columns. This way we could avoid doing:
file path => load file => write file => load file using the given preprocessor => predict.